server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -417,81 +467,37 @@ private void loadAllLookupsAndInitStateRef()
-      if (lookupBeanList == null) {
-        LOG.info("Coordinator is unavailable. Loading saved snapshot instead");
-        lookupBeanList = getLookupListFromSnapshot();
-      }
-    } else {
-      lookupBeanList = getLookupListFromSnapshot();
-    }
-    return lookupBeanList;
-  }
-
   private void startLookups(List<LookupBean> lookupBeanList)
   {
     ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
     ExecutorService executorService = Execs.multiThreaded(
         lookupConfig.getNumLookupLoadingThreads(),
         "LookupReferencesManager-Startup-%s"
     );
-    ExecutorCompletionService completionService = new ExecutorCompletionService(executorService);
-    List<Future<Map.Entry>> futures = new ArrayList<>();
+    CompletionService<Map.Entry<LookupBean, LookupExtractorFactoryContainer>> completionService =
+        new ExecutorCompletionService<>(executorService);
+    List<Future<?>> allFutures = new ArrayList<>();
     try {
       LOG.info("Starting lookup loading process");
-      for (LookupBean lookupBean : lookupBeanList) {
-        futures.add(
-            completionService.submit(
-                () -> {
-                  LookupExtractorFactoryContainer container = lookupBean.getContainer();
-                  LOG.info(
-                      "Starting lookup [%s]:[%s]",
-                      lookupBean.getName(),
-                      container
-                  );
-                  if (container.getLookupExtractorFactory().start()) {
-                    LOG.info(
-                        "Started lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
-                  } else {
-                    LOG.error(
-                        "Failed to start lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return null;
-                  }
-                }
-            )
-        );
-      }
-      for (int i = 0; i < futures.size(); i++) {
-        try {
-          final Future<AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer>> completedFuture = completionService
-              .take();
-          final AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer> lookupResult = completedFuture
-              .get();
-          if (lookupResult != null) {
-            builder.put(lookupResult);
-          }
-        }
-        catch (ExecutionException e) {
-          LOG.error(e, "Execution error during lookup loading.");
+      for (int i = 0; i < lookupConfig.getLookupStartRetries(); i++) {
+        LOG.info("Round of attempts #%d, [%d] lookups", i + 1, lookupBeanList.size());
+        Pair<Map<LookupBean, LookupExtractorFactoryContainer>, List<LookupBean>> results =
```

Comment:
Can this just return a `Map<String, LookupExtractorFactoryContainer>` which contains the successes where the key is the name?

The failures can be deduced by the difference between `lookupBeanList` and the names which are not in the return value.

Reply:
My concern was that multiple lookups could have the same name, or it's not possible?

Reply:
At the very least it would break the immutable map builder.

I don't think different names should be able to come in at the same time.

Things like `io.druid.query.lookup.LookupReferencesManager.LookupUpdateState` also assume a string to `LookupExtractorFactoryContainer` map, so I think if a lookup came in somehow with the name more than once it could break things.

Reply:
Ok, changed